core/templates/filters/translate.pipe.ts

```diff
@@ -0,0 +1,149 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Translate pipe for i18n translations.
+ */
+
+import { ChangeDetectorRef, OnDestroy, Pipe, PipeTransform }
+  from '@angular/core';
+import { Subscription } from 'rxjs';
+import { LangChangeEvent, TranslateService } from 'services/translate.service';
+import { UtilsService } from 'services/utils.service';
+
+/**
+ * Commonly used terms in this file.
+ * key - Key for i18n translation
+ * interpolateParams or params - Params for interpolation
+ * interpolatedValue - The final translation that is returned.
+ * Note: Intentionally left out the L of innerHTM"L" (only in this file) to
+ * avoid the linting errors.
+ * Example: <h1 [innerHTM]="'I18N_ABOUT_PAGE_HEADING' | translate:{x: 'val'}">
+ * 'I18N_ABOUT_PAGE_HEADING' is referred here as key.
+ * "translate" is the pipe. Every pipe must have a transform function. The
+ * transform function is called when angular encouters the pipe in HTML.
+ * The object following the pipe, i.e. {x: 'val'}, is another argument to the
+ * transform function. This object is called params or interpolationParams.
+ * Note: For every "| translate" angular encounters in the html, it creates a
+ * seperate instance of this pipe.
+ */
+
+@Pipe({
+  name: 'translate',
+  // This option is required to update the interpolatedValue when promises are
+  // resolved.
+  pure: false
+})
+export class TranslatePipe implements PipeTransform, OnDestroy {
+  interpolatedValue: string = '';
+  /**
+   * Variable to keep track of the last key sent for translation. If the
+   * previously sent key and params are same as the current one, the pipe
+   * doesn't request for translation and interpolation again.
+   */
+  lastKey: string;
+  /**
+   * Variable to keep track of the last param sent for translation. If the
+   * previously sent key and params are same as the current one, the pipe
+   * doesn't request for translation and interpolation again.
+   */
+  lastParams: Object;
+  onLangChange: Subscription;
+
+  constructor(
+    private translateService: TranslateService,
+    private ref: ChangeDetectorRef,
+    private utilsService: UtilsService
+  ) {}
+
+  /**
+   * Updates the value of interpolatedValue.
+   * @param {string} key - key for i18n translation
+   * @param {Object} interpolateParams - Params for interpolation
+   */
+  updateInterpolatedValue(key: string, interpolateParams?: Object): void {
+    this.interpolatedValue = this.translateService.getInterpolatedString(
+      key, interpolateParams);
+
+    // Storing the key to check if the key is same when the transform is invoked
+    // again.
+    this.lastKey = key;
+
+    // Storing the params to check if the params are same when the transform is
+    // invoked again.
+    this.lastParams = interpolateParams;
+    this.ref.markForCheck();
+  }
+
+  /**
+   * @param {string} key - key for i18n
+   * @param {Object} params - params for interpolation
```

Comment:
Typedef here does not match typedef in function signature. If undefined is allowed then explain its significance here.

Also, it's OK to have longer arg descriptions if you need to add more explanatory detail for things. Try not to define things recursively, in terms of themselves. For example, I wouldn't use "params" to define "params". I would say something like "A dict of key-value pairs. The keys represent ... and the corresponding values represent ...". Remember that the reader doesn't know what you already know, so err on the side of being descriptive.